app/app/dialog/[id]/page.tsx

```diff
@@ -1,16 +1,241 @@
+      </div>
+    </div>
+  )
+}
+
+const getFromStoracha = async (cid: string) => {
+  const url = new URL(`/ipfs/${cid}`, process.env.NEXT_PUBLIC_STORACHA_GATEWAY_URL)
+  const response = await fetch(url)
+  if (!response.ok) {
+    throw new Error(`Failed to fetch: ${response.status} ${response.statusText} ${url}`);
+  }
+  return response
+}
+
 export default function Page () {
+  const [{ client }] = useTelegram()
+  const [loading, setLoading] = useState(true)
+  const [dialogData, setDialogData] = useState<DialogData | null>(null)
+  const [messages, setMessages] = useState<MessageData[]>([])
+  const [userId, setUserId] = useState<string>()
+  const [participants, setParticipants] = useState<Record<string, EntityData>>({})
   const params = useParams<{ id: string }>()
-  console.log(params)
+  const searchParams = useSearchParams()
+  const backupCid = searchParams.get('backupData')
+
+  useEffect(() => {
+      const restoreBackup = async () => {
+        try{
+          if (!client.connected) await client.connect()
+
+          const userId = (await client.getMe()).id
+          setUserId(userId.toString())
+
+          if(!cloudStorage.getKeys.isAvailable()){
+            throw new Error('Error trying to access cloud storage')
+          }
+          
+          const encryptionPassword = await cloudStorage.getItem('encryption-password')
+          if (!encryptionPassword) {
+            throw new Error('Encryption password not found in cloud storage')
+          }
+          
+          const response = await getFromStoracha(`${backupCid}?format=car`)
+          const car = new Uint8Array(await response.arrayBuffer());
+          console.log('CAR file fetched')
+
+          // decode
+          const { roots } = CAR.decode(car)
+          if (!roots.length) {
+            throw new Error('Missing root block')
+          }
+          const { code } = roots[0].cid
+          if (code !== dagCBOR.code) {
+          throw new Error(`unexpected root CID codec: 0x${code.toString(16)}`)
+          }
+
+          console.log('decoding...')
+          const root = roots[0]
+          const backupRaw = dagCBOR.decode(root.bytes) as BackupModel
+          console.log('Decoded CAR root')
+          // console.log(JSON.stringify(backupRaw))
+
+          const id = params.id.replace('-', '')
+          const dialogCid = backupRaw['tg-miniapp-backup@0.0.1'].dialogs[id].toString()
+          const dialogResult = await getFromStoracha(dialogCid)
+          const encryptedDialogData = new Uint8Array(await dialogResult.arrayBuffer())
+
+          const decryptedDialogData = dagCBOR.decode(
+            await Crypto.decryptContent(encryptedDialogData, encryptionPassword)
+          ) as DialogData
+          
+          console.log(JSON.stringify(decryptedDialogData))
+       
+          // Restore messages and entities
+          const restoredMessages: MessageData[] = [];
+          for (const messageLink of decryptedDialogData.messages) {
```

Comment:
For a future PR - I'd avoid buffereing the entire chat history in memory. Instead, just get the first chunk and retrieve more when/if the user scrolls to the top of the history. Remember this is max 1,000 messages so quite big.